### PR TITLE
[core][metric] Add fallback memory and seal memory breakdown 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

#29485 with mac tests skipped (they are just slow) 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 1]
```diff
@@ -72,6 +90,20 @@ void ObjectStatsCollector::OnObjectSealed(const LocalObject &obj) {
...
+    counter_type = obj.Sealed()
+                       ? std::make_pair(/* fallback_allocated */ true, /* sealed*/ true)
+                       : std::make_pair(/* fallback_allocated */ true, /* sealed*/ false);
+  } else {
+    counter_type =
+        obj.Sealed() ? std::make_pair(/* fallback_allocated */ false, /* sealed*/ true)
+                     : std::make_pair(/* fallback_allocated */ false, /* sealed*/ false);
+  }
+
+  bytes_by_loc_seal_.Decrement(counter_type, kObjectSize);
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:106]:
```suggestion
  bytes_by_loc_seal_.Decrement(std::make_pair(kAllocation.fallback_allocated, obj.Sealed()), kObjectSize);
```

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 2]
```diff
@@ -176,16 +208,29 @@ int64_t ObjectStatsCollector::GetNumBytesCreatedCurrent() const {
...
+      {{ray::stats::LocationKey.name(), ray::stats::kObjectLocMmapShm},
+       {ray::stats::ObjectStateKey.name(), ray::stats::kObjectUnsealed}});
 
+  // Fallback memory sealed
   ray::stats::STATS_object_store_memory.Record(
-      num_bytes_unsealed_,
-      {{ray::stats::LocationKey.name(), ray::stats::kObjectLocUnsealed}});
+      bytes_by_loc_seal_.Get({/* fallback_allocated */ true, /* sealed */ true}),
+      {{ray::stats::LocationKey.name(), ray::stats::kObjectLocMmapDisk},
+       {ray::stats::ObjectStateKey.name(), ray::stats::kObjectSealed}});
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:227]: I think you don't need the `.name()` here.

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 3]
```diff
@@ -24,10 +24,17 @@ namespace plasma {
 void ObjectStatsCollector::OnObjectCreated(const LocalObject &obj) {
   const auto kObjectSize = obj.GetObjectInfo().GetObjectSize();
   const auto kSource = obj.GetSource();
+  const auto &kAllocation = obj.GetAllocation();
+
+  if (kAllocation.fallback_allocated) {
+    bytes_by_loc_seal_.Increment({/*fallback_allocated*/ true, /*sealed*/ false},
+                                 kObjectSize);
+  } else {
+    bytes_by_loc_seal_.Increment({/*fallback_allocated*/ false, /*sealed*/ false},
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:33]:
```suggestion
    bytes_by_loc_seal_.Increment({kAllocation.fallback_allocated, /*sealed*/ false},
```

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 4]
```diff
@@ -52,6 +59,17 @@ void ObjectStatsCollector::OnObjectSealed(const LocalObject &obj) {
   RAY_CHECK(obj.Sealed());
   const auto kObjectSize = obj.GetObjectInfo().GetObjectSize();
 
+  const auto &kAllocation = obj.GetAllocation();
+  if (kAllocation.fallback_allocated) {
+    bytes_by_loc_seal_.Swap({/*fallback_allocated*/ true, /* sealed */ false},
+                            {/*fallback_allocated*/ true, /* sealed */ true},
+                            kObjectSize);
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:66]: Same here?

File: src/ray/object_manager/plasma/stats_collector.cc (hunk 3 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/stats_collector.cc:33]: oh lol - somehow didn't realize it. Good catch.

File: src/ray/object_manager/plasma/stats_collector.cc (hunk 2 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/stats_collector.cc]: Not with the current `Record` but add another one that skips a bunch of tag key register operations.

**Review by ericl — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**